### DISCUSSION TITLE
Reverse Proxy middleware

### DISCUSSION THREAD

**Discussion by simplify-web** [pocketbase/pocketbase#1013] [Q&A]
Hi,
thanks for your great project. Golang noob here. 
I'd like to add a reverse proxy to pocketbase. The aim is to proxy /db to a couchdb server, where every pocketbase user would get their own database depending on their pocketbase _id.
I managed to proxy a single path by adding to a custom main.go:
```		
// create the reverse proxy
url, _ := url.Parse("http://localhost:5984")
proxy := httputil.NewSingleHostReverseProxy(url)
e.Router.Any("/db", echo.WrapHandler(proxy))
```
What I would need is a wildcard match and a rewrite. so that every URI from /db/userdb/* is proxied to localhost:5984/userdb-USERID/* . 
The rewrite should only be done after checking that the user is authenticated.

Any help would be highly appreciated
Thanks

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand the question, but you can have a wildcard portion in your route path by appending `*`, eg.:

```go
e.Router.Any("/db/*", echo.WrapHandler(proxy))
```
_(the matched wildcard path will be available as regular Path parameter in the echo.Context of your handler)_

**Reply by simplify-web** [reply]
Thanks for your quick answer.
This works well for the wildcard part. 
So now I only need an option to get the user id and rewrite the destination, e.g.
/db/userdb/* is proxied to localhost:5984/userdb-ID/*
This can probably only be done in a hook for every request.
Any ideas how to achieve this?
Thanks again.

**Reply by ganigeorgiev** [reply]
You can get the authenticated user from the echo.Context store in your handler like this:

```go
func(c echo.Context) error {
    // for PocketBase v0.7.x
    loggedUser, _ := c.Get("user").(*models.User) // you can also use the key const apis.ContextUserKey

    // for PocketBase v0.8.0-rc    
    loggedUser, _ := c.Get("authRecord").(*models.Record) // you can also use the key const apis.ContextAuthRecordKey
    
    if loggedUser != nil {
        print(loggedUser.Id)
    }
    
    ...
}
```

**Reply by simplify-web** [reply]
I am not sure how I would incorporate your code @ganigeorgiev into my proxy. I have a working proxy now which creates a proxy and rewrites the path depending on the userId, which I get from the Header
``` go
package main

import (
	"log"
	"net/http"
	"net/http/httputil"
	"strings"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		director := func(req *http.Request) {
			userId := req.Header.Get("userId")
			urlPath := strings.Replace(req.URL.Path, "userdb", "userdb-"+userId, 1)
			req.URL.Scheme = "http"
			req.URL.Host = "localhost:5984"
			req.URL.Path = urlPath
		}

		// create the reverse proxy
		proxy := &httputil.ReverseProxy{Director: director}
		e.Router.Any("/userdb/*", echo.WrapHandler(proxy))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

This is of course not safe, because anybody could fake the headers and be redirected to a different couchdb. So I either find a way to access the authContext or get the userId from the JWT, which I would need to check against the JWT secret key. As far as I understand the docs, there is not one JWT secret for all users which I could use. (Its created from two parts) .
So I'm quite stuck here.
Thanks again for your time.

**Reply by ganigeorgiev** [reply]
@simplify-web I'm not sure that I fully understand the above example, but my idea was to replace the default `echo.WrapHandler` with a custom one and call the proxy inside it (at the end WrapHandler doesn't do anything else either). Or in other words, the above example could look like this:

```go
e.Router.Any("/userdb/*", func(c echo.Context) error {
	// for PocketBase v0.7.x
	loggedUser, _ := c.Get("user").(*models.User) // you can also use the key const apis.ContextUserKey

	// for PocketBase v0.8.0-rc
	// loggedUser, _ := c.Get("authRecord").(*models.Record) // you can also use the key const apis.ContextAuthRecordKey

	if loggedUser == nil {
		// for PocketBase v0.7.x
		return rest.NewForbiddenError("Only authorized users can access this.", nil)

		// for PocketBase v0.8.0-rc
		// return apis.NewForbiddenError("Only authorized users can access this.", nil)
	}

	proxy := &httputil.ReverseProxy{
		Director: func(req *http.Request) {
			urlPath := strings.Replace(req.URL.Path, "userdb", "userdb-"+loggedUser.Id, 1)
			req.URL.Scheme = "http"
			req.URL.Host = "localhost:5984"
			req.URL.Path = urlPath
		},
	}

	proxy.ServeHTTP(c.Response(), c.Request())

	return nil
})
```

**Reply by simplify-web** [reply]
Thank you @ganigeorgiev. very elegant. now I only need the missing part to get the proper client request, because I will not be able to use the js-sdk. My App will be offine first and the connection to couchdb will be with pouchdb for the syncing. 
I tried to use the jwt from the browser after authenticating with the js-sdk but i cant get "authRecord" from your code example. Its always nil. Should the curl command below even work?
`curl -H 'Accept: application/json'  -H "Authorization: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2Njk2MzE0MjYsImlkIjoiY29qdWpxbzd5MzlxMHJkIiwidHlwZSI6ImFkbWluIn0.F0j3k8mZAN52qv-lh16ARHNmdH7jG9YTfh9M-RF0U3A" http://localhost:8090/userdb/3`

Also I*m using pocketbase [v0.8.0-rc2](https://github.com/pocketbase/pocketbase/releases/tag/v0.8.0-rc2). So my question is the userId really called authRecord in this version?

Any more advice how to make this final step work, would be greatly appreciated.
Big thank you

**Reply by ganigeorgiev** [reply]
@simplify-web The payload of the attached token is for admin, not an auth record. Make sure that you are sending an app "users" token (or whatever else your auth collection is named). Or in other words, in the above curl request you are authenticating as admin, not an auth record.

**Reply by simplify-web** [reply]
Your really great Gani. Thanks so much. 
One more question. Are the user ids always ascii without any special chars?
Where do you need the most help in the js-sdk? That's something I could help you.
Have a great day
Andrej

**Reply by ganigeorgiev** [reply]
>  Are the user ids always ascii without any special chars?

By default the ids are auto generated based on the following alphabet:

```
abcdefghijklmnopqrstuvwxyz0123456789
```
But please note that users are allowed to submit their own ids, 15 chars string (in order to support offline-first apps), so if your user registration is public then it is possible that someone could submit a 15 chars id that contains other symbols. The only requirement is to be compatible with the following regex:

```
^[^\@\#\$\&\|\.\,\'\"\\\/\s]+$
```

If you want to prevent non-ascii ids submissions (or to further "sanitize" them), you can attach for example to the `OnRecordBeforeCreateRequest` and change/extra-validate the id from there:

```go
app.OnRecordBeforeCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	if e.Record.Collection().Name != "users" {
		return nil // not a user create
	}

	// do something with the id OR always overwrite it to auto generated one
	// print(e.Recrod.Id)
	e.Record.RefreshId()

	return nil
})
```

---

> Where do you need the most help in the js-sdk?

Currently there are no critical open issues with the js-sdk. Only documentation and better examples with various frontend frameworks is missing (eg. nextjs13) but it may be better such guides to be part of the site documentation for better visibility (with the v0.8.0 release the site docs will be open-sourced so external contributions will be possible).
